shoulder/generator/c_header_generator.py

```diff
@@ -21,463 +21,767 @@
+                        self._generate_bitfield_is_set_val(outfile, reg, field)
+                        self._generate_bitfield_clear(outfile, reg, field)
+                        self._generate_bitfield_clear_val(outfile, reg, field)
+                        self._generate_bitfield_is_clear(outfile, reg, field)
+                        self._generate_bitfield_is_clear_val(outfile, reg, field)
+                    else:
+                        self._generate_field_get(outfile, reg, field)
+                        self._generate_field_get_val(outfile, reg, field)
+                        self._generate_field_set(outfile, reg, field)
+                        self._generate_field_set_val(outfile, reg, field)
 
     def _generate_register_comment(self, outfile, reg):
-        reg_comment = "// {name} ({long_name})\n// {purpose}\n".format(
-            name = str(reg.name),
-            long_name = str(reg.long_name),
-            purpose = str(reg.purpose)
+        comment = "// {name} ({long_name})\n".format(
+            name=str(reg.name),
+            long_name=str(reg.long_name)
         )
-        outfile.write(reg_comment)
+        outfile.write(comment)
+
+        comment = str(reg.purpose)
+        wrapped = textwrap.wrap(comment, width=75)
+        for line in wrapped:
+            line = "// " + str(line) + "\n"
+            outfile.write(line)
 
     def _generate_fieldset_comment(self, outfile, fieldset):
-        if(fieldset.condition is not None):
-            fieldset_comment = "\t// Fieldset valid when: {comment}\n".format(
-                    comment = str(fieldset.condition)
-                )
-            outfile.write(fieldset_comment)
-
-    def _generate_sysreg_get(self, outfile, reg):
-        access_name = str(reg.access_attributes.mnemonic).lower()
+        if fieldset.condition is not None:
+            fieldset_comment = "Fieldset valid when: {comment}\n".format(
+                comment=str(fieldset.condition))
+            wrapped = textwrap.wrap(fieldset_comment, width=71)
+            for line in wrapped:
+                line = "\t// " + str(line) + "\n"
+                outfile.write(line)
+
+# ----------------------------------------------------------------------------
+# register_get
+# ----------------------------------------------------------------------------
+
+    def _generate_register_get(self, outfile, reg):
+        """
+        Generate a C function that reads the given register
+        """
+
+        rname = reg.name.lower()
+        prefix = self._register_function_prefix(reg)
+        suffix = config.register_read_function
+        if reg.size == 32:
+            size_type = "uint32_t"
+        else:
+            size_type = "uint64_t"
+
+        gadget = self.gadgets["shoulder.c.function_definition"]
+        gadget.name = prefix + "_" + rname + "_" + suffix
+        gadget.return_type = size_type
+        gadget.args = []
+
+        if reg.access_mechanisms["mrs_register"]:
```

Comment:
So this will take only one mechanism to generate even if the register has multiple mechanisms of the same type (such as two MRS mechanisms) or multiple types (such as an MRS and a MRS_Banked). I think it would be confusing if we had a bunch of different generated functions that do the same thing, but do you think we are losing info or tests by omitting the others? Is there a way we should decide which one to choose if we only select one?

Reply:
This is a very good point, and one that I don't really have an answer for yet. I honestly don't know all of the intricacies of choosing the right access mechanism from every context you can read a register from. I went with the approach of "find the first one that will work, and use that".

I think this approach supports the more general-purpose use case that the base Shoulder project is trying to provide (creating a general-purpose register access API). Maybe we can play around with choosing access mechanisms differently in Scapula specific generators?

Reply:
Yeah, I've also been trying to think of how to go about this and I haven't come up with anything that I like. The best I can think of would be like having a whole bunch of `get()` functions like: `aarch64_my_register_get_mrs_1`, `aarch64_my_register_get_mrs_2`, `aarch64_my_register_get_mrc` but I really don't like that idea and I think it really clutters up the generated shoulder. I too don't know how the intricacies work and I imagine that a lot of it is register specific with weird corner cases all over the place so finding a universal solution doesn't seem very feasible.

I think this approach works best unless something else jumps out at us. Like I said, I think the first mechanisms described in the XML are the most commonly used ones, so them being used first is probably best.

Like I said, I agree with how you did this. I was just curious to what thoughts you might have since I want to try and get as accurate and in-depth info as I can into the registers. I think you're right though, for this general purpose register API, we should use the basic accessors and then in Scapula we can do more varied and unique tests.